library/src/main/java/com/github/pwittchen/reactivewifi/ReactiveWifi.java

```diff
@@ -125,6 +127,78 @@
   }
 
+  /**
+   * Observes the current WPA supplicant state.
+   * Returns the current WPA supplicant as a member of the {@link SupplicantState} enumeration,
+   * returning {@link SupplicantState#UNINITIALIZED} if Wifi is not enabled.
+   *
+   * @param context Context of the activity or an application
+   * @return RxJava Observable with SupplicantState
+   */
+  public Observable<SupplicantState> observeSupplicantState(final Context context) {
+    final IntentFilter filter = new IntentFilter();
+    filter.addAction(WifiManager.SUPPLICANT_STATE_CHANGED_ACTION);
+
+    return Observable.create(new Observable.OnSubscribe<SupplicantState>() {
+      @Override public void call(final Subscriber<? super SupplicantState> subscriber) {
+        final BroadcastReceiver receiver = new BroadcastReceiver() {
+          @Override public void onReceive(Context context, Intent intent) {
+            SupplicantState supplicantState =
+                intent.getParcelableExtra(WifiManager.EXTRA_NEW_STATE);
+
+            if (SupplicantState.isValidState(supplicantState)) {
+               subscriber.onNext(supplicantState);
+            }
+          }
+        };
+
+        context.registerReceiver(receiver, filter);
+
+        subscriber.add(unsubscribeInUiThread(new Action0() {
+          @Override public void call() {
+            context.unregisterReceiver(receiver);
+          }
+        }));
+      }
+    }).defaultIfEmpty(SupplicantState.UNINITIALIZED);
+  }
+
+  /**
+   * Observes the Wifi network the device is connected to.
+   * Returns the current Wifi network information as a {@link WifiInfo} object, or {@code null}
+   * if Wifi is not enabled.
+   *
+   * @param context Context of the activity or an application
+   * @return RxJava Observable with WifiInfo
+   */
+  public Observable<WifiInfo> observeWifiAccessPointChanges(final Context context) {
+    final WifiManager wifiManager = (WifiManager) context.getSystemService(Context.WIFI_SERVICE);
+    final IntentFilter filter = new IntentFilter();
+    filter.addAction(WifiManager.SUPPLICANT_STATE_CHANGED_ACTION);
+
+    return Observable.create(new Observable.OnSubscribe<WifiInfo>() {
+      @Override public void call(final Subscriber<? super WifiInfo> subscriber) {
+        final BroadcastReceiver receiver = new BroadcastReceiver() {
+          @Override public void onReceive(Context context, Intent intent) {
+            SupplicantState supplicantState =
+                    intent.getParcelableExtra(WifiManager.EXTRA_NEW_STATE);
+            if (supplicantState == SupplicantState.COMPLETED) {
+              subscriber.onNext(wifiManager.getConnectionInfo());
+            }
+          }
+        };
+
+        context.registerReceiver(receiver, filter);
+
+        subscriber.add(unsubscribeInUiThread(new Action0() {
+          @Override public void call() {
+            context.unregisterReceiver(receiver);
+          }
+        }));
+      }
+    }).defaultIfEmpty(null);
```

Comment:
I think, returning null here is not a good idea. I'd rather avoid passing and returning nulls when possible. We can remove `defaultIfEmpty(null)` instruction and it will be fine.